### PR TITLE
iox-1613 replace EXPECT_DEATH with custom mechanism

### LINKED ISSUE #1613: Remove `EXPECT_DEATH`
## Required information

In our cases using `EXPECT_DEATH` come often with undefined behavior which is unacceptable for certified tests. Possible approaches are:

* Create thread and call `pthread_exit` inside the error handler, would require the implementation of that feature in #1365 
* Throw an exception in the error handler, would require heavy refactoring and feature #1392 
* Another nifty idea.

## ToDo
- [x] (high prio) Define an alternative technique to `EXPECT_DEATH` and document it in the iceoryx testing document
- [x] (high prio) Remove all occurrences of `EXPECT_DEATH` in cxx tests
- [x] (medium prio) Remove all occurrences of `EXPECT_DEATH` in remaining hoofs
- [ ] (low prio) Remove all occurrences of `EXPECT_DEATH` in the remaining iceoryx repo
  - Using `EXPECT_FATAL_FAILURE` in combination with the `RouDiEnvironment` results in triggering the leak sanitizer
  - The same is the case when the SUT allocate in the call which leads to the fatal failure, e.g. the ctor of `FileReader`

Can be split up in further issues when required.

Extracted from: #1196

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

This PR adds a replacement for `EXPECT_DEATH` which is based on `longjmp` and replaces the usage in `iceoryx_hoofs`.

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1613

### CODE REVIEW

File: iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/fatal_failure.hpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            iox::ErrorHandlerMock::setTemporaryErrorHandler<ErrorType>([&](const auto error, const auto errorLevel) {
+                EXPECT_THAT(error, ::testing::Eq(expectedError));
+                EXPECT_THAT(errorLevel, ::testing::Eq(expectedErrorLevel));
+
+                // NOLINTNEXTLINE(cert-err52-cpp) exception cannot be used and longjmp/setjmp is a working fallback
+                std::longjmp(&jmpBuffer[0], JMP_VALUE);
+            });
+
+        // NOLINTNEXTLINE(cert-err52-cpp) exception cannot be used and longjmp/setjmp is a working fallback
+        if (setjmp(&jmpBuffer[0]) == JMP_VALUE)
```
Comment (elBoberido) [inline iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/fatal_failure.hpp:46]: @MatthiasKillat it seems to work. Thanks for the hint

**Review by FerdinandSpitzschnueffler — APPROVED** [review]

**Review by MatthiasKillat — APPROVED** [review]
Looks good. Only minor observations about lifetime of ad hoc functions. The remainder is more about what should be a fatal error and checked in a test, but this is not to be addressed here.

File: doc/design/error-handling.md [hunk 2]
```diff
@@ -264,6 +264,23 @@ auto errorFunc = [](Error& error) {
...
+The `EXPECT_DEATH` gTest macro forks the process which slows down the test execution (especially with the ThreadSanitizer enabled)
+and causes issues with running thread. The `IOX_EXPECT_FATAL_FAILURE` registers a temporary error handler and runs the provided
+function in a separate thread. When the error handler is called `longjmp` is used to prevent the termination and instead ensures
+to gracefully shutdown the thread.
+
+```cpp
+#include "iceoryx_hoofs/testing/fatal_failure.hpp"
+TEST(MyTest, valueOnNulloptIsFatal) {
+    iox::optional<bool> sut;
+    IOX_EXPECT_FATAL_FAILURE<iox::HoofsError>([&] { sut.value(); }, iox::HoofsError::EXPECTS_ENSURES_FAILED);
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply doc/design/error-handling.md:280]: @MatthiasKillat yes, it expects a `std::function<void()>` and `std::function<void(ErrorType)>`. This can be changed with the new error reporting

File: iceoryx_binding_c/test/moduletests/test_runtime.cpp [hunk 3]
```diff
@@ -69,13 +72,19 @@ TEST_F(BindingC_Runtime_test, RuntimeNameLengthIsOutOfLimit)
     ::testing::Test::RecordProperty("TEST_ID", "8fd6735d-f331-4c9c-9a91-3f06d3856d15");
     std::string tooLongName(iox::MAX_RUNTIME_NAME_LENGTH + 1, 's');
 
-    EXPECT_DEATH({ iox_runtime_init(tooLongName.c_str()); }, ".*");
+    IOX_EXPECT_FATAL_FAILURE<iox::HoofsError>(
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_binding_c/test/moduletests/test_runtime.cpp:75]: @MatthiasKillat yes, has to be discussed but we somehow have to check if this edge case is handled

File: iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/fatal_failure.hpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+/// @brief This function is the base for 'IOX_EXPECT_FATAL_FAILURE' and 'IOX_EXPECT_NO_FATAL_FAILURE' and should not be
+/// used by its own. The function only works in combination with the iceoryx error handler.
+/// @tparam[in] ErrorType The error type which is expected, e.g. 'iox::HoofsError'
+/// @param[in] testFunction This function will be executed as SUT and might call the error handler with a 'FATAL' error
+/// level
+/// @param[in] onFatalFailurePath This function will be executed on the failure path after the failure was detected
+/// @param[in] onNonFatalFailurePath This function will be executed on the non-failure path if no failure was detected
+/// @return true if a fatal failure occurs, false otherwise
+template <typename ErrorType>
+bool IOX_FATAL_FAILURE_TEST(const std::function<void()>& testFunction,
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/fatal_failure.hpp:43]: clang-tidy suggested to use reference so I could either use a reference or add a suppression ;)

File: iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/fatal_failure.inl [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                std::longjmp(&jmpBuffer[0], JMP_VALUE);
+            });
+
+        // NOLINTNEXTLINE(cert-err52-cpp) exception cannot be used and longjmp/setjmp is a working fallback
+        if (setjmp(&jmpBuffer[0]) == JMP_VALUE)
+        {
+            hasFatalFailure = true;
+            // using value directly is save since this path is only executed if the error handler was called and the
+            // respective values were set
+            onFatalFailurePath(detectedError.value(), detectedErrorLevel.value());
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/testing/include/iceoryx_hoofs/testing/fatal_failure.inl:57]: @MatthiasKillat we can change this with the new error reporting
